### PR TITLE
feat: fetch rss command

### LINKED ISSUE #11: new command fetch_rss to add torrents
Could be used together with a cron job.
Transmission doesn't seem to support it natively

### PR DESCRIPTION
**Pull request by aidanaden**
closes #11 

also includes regex filtering (use claude.ai for regex help)

### CODE REVIEW

File: rm-main/src/cli.rs [hunk 1]
```diff
@@ -60,3 +63,56 @@ async fn add_torrent(config: &Config, torrent: String) -> Result<()> {
...
+                    println!("{title} does not match provided regex");
+                }
+            } else {
+                return item.link();
+            }
+        }
+        None
+    });
+    for url in urls {
+        println!("downloading {url}");
```
Comment (aidanaden) [inline rm-main/src/cli.rs:103]: placeholder, not sure on what exactly to print. maybe include name?

Comment (micielski) [reply rm-main/src/cli.rs:103]: Yeah, maybe just the name would be better?

File: Cargo.toml [hunk 2]
```diff
@@ -28,6 +28,7 @@ url = "2.5"
 toml = "0.8"
 rss = "2.0.8"
 reqwest = "0.12.5"
+regex = "1.10.5"
```
Comment (micielski) [inline Cargo.toml:31]: regex = "1"

**Review by micielski — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 3]
```diff
@@ -26,6 +26,9 @@ base64 = "0.22"
 xdg = "2.5"
 url = "2.5"
 toml = "0.8"
+rss = "2.0.8"
+reqwest = "0.12.5"
```
Comment (micielski) [inline Cargo.toml:30]: reqwest = "0.12"

File: Cargo.toml [hunk 4]
```diff
@@ -26,6 +26,9 @@ base64 = "0.22"
 xdg = "2.5"
 url = "2.5"
 toml = "0.8"
+rss = "2.0.8"
```
Comment (micielski) [inline Cargo.toml:29]: rss = "2"

File: rm-main/src/cli.rs [hunk 5]
```diff
@@ -60,3 +63,56 @@ async fn add_torrent(config: &Config, torrent: String) -> Result<()> {
...
+        let args = TorrentAddArgs {
+            filename: Some(url.to_string()),
+            ..Default::default()
+        };
+        if let Err(e) = transclient.torrent_add(args).await {
+            eprintln!("error while adding a torrent: {e}");
+            if e.to_string().contains("expected value at line") {
+                eprintln!("Check whether your arguments are valid.");
```
Comment (micielski) [inline rm-main/src/cli.rs:111]: Anyhow could be used here also

**Review by micielski — CHANGES_REQUESTED** [review]

File: rm-main/src/cli.rs [hunk 6]
```diff
@@ -63,12 +64,35 @@ async fn add_torrent(config: &Config, torrent: String) -> Result<()> {
...
 }
 
-async fn fetch_rss(config: &Config, url: String) -> Result<()> {
+async fn fetch_rss(config: &Config, url: &str, filter: &str) -> Result<()> {
     let mut transclient = transmission::utils::client_from_config(&config);
     let content = reqwest::get(url).await?.bytes().await?;
     let channel = rss::Channel::read_from(&content[..])?;
-    let urls = channel.items().iter().filter_map(|item| item.link());
+    let re = {
+        let re = Regex::new(&format!(r"{filter}"));
```
[the remark this answers is not in the document]
Comment (aidanaden) [reply rm-main/src/cli.rs]: the filter is a regex so using raw strings makes things easier i assume - i was just following the example given https://docs.rs/regex/latest/regex/#examples

[the remark this answers is not in the document]
Comment (micielski) [reply rm-main/src/cli.rs]: Raw strings are for programmers, not users, to just be able to put quotes, backslashes without any escaping

[the remark this answers is not in the document]
Comment (aidanaden) [reply rm-main/src/cli.rs]: will backslashes and other special characters in the user's regex string need to be escaped if we don't use raw strings? 

if escaping is not required for the regex string provided by the user i'll remove the raw string

**Comment by aidanaden** [comment]
any specific reason for the specific crate versions?

**Comment by micielski** [comment]
> any specific reason for the specific crate versions?

There will be no need to manually bump them in Cargo.toml, `cargo update` will be enough and nothing will break because they're stable and use [semantic versioning](https://semver.org/)

File: rm-main/src/cli.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (micielski) [reply rm-main/src/cli.rs]: You can safely get rid of this `r` before the string. It will still work the same way for the user

**Pull request merged**